pkg/resource/deployment_test.go

```diff
@@ -0,0 +1,220 @@
+)
+
+func TestChecksum(t *testing.T) {
+	tests := []struct {
+		name          string
+		origSecret    *corev1.Secret
+		updatedSecret *corev1.Secret
+		validate      func(*testing.T, string, *appsapi.Deployment)
+	}{
+		{
+			name: "no secret changes",
+			origSecret: testSecret(map[string][]byte{
+				"credentials": []byte("orig creds"),
+			}),
+			validate: func(t *testing.T, origHash string, dep *appsapi.Deployment) {
+				currentHash := dep.Annotations[defaults.ChecksumOperatorAnnotation]
+				if origHash != currentHash {
+					t.Errorf("Has unexpectedly changed from %s to %s", origHash, currentHash)
+				}
+			},
+		},
+		{
+			name: "secret changes",
+			origSecret: testSecret(map[string][]byte{
+				"credentials": []byte("orig creds"),
+			}),
+			updatedSecret: testSecret(map[string][]byte{
+				"credentials": []byte("updated creds"),
+			}),
+			validate: func(t *testing.T, origHash string, dep *appsapi.Deployment) {
+				currentHash := dep.Annotations[defaults.ChecksumOperatorAnnotation]
+				if origHash == currentHash {
+					t.Errorf("Hash unexpectedly didn't change from %s", origHash)
+				}
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			annotatedNamespace := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: defaults.ImageRegistryOperatorNamespace,
+					Annotations: map[string]string{
+						defaults.SupplementalGroupsAnnotation: "1/2",
+					},
+				},
+			}
+
+			imageRegistryCertificatesConfigMap := &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      defaults.ImageRegistryCertificatesName,
+					Namespace: defaults.ImageRegistryOperatorNamespace,
+				},
+			}
+
+			kubeClient := fake.NewSimpleClientset(test.origSecret, annotatedNamespace, imageRegistryCertificatesConfigMap)
+			kubeInformer := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+
+			configClient := fakeconfig.NewSimpleClientset()
+			configInformer := configinformers.NewSharedInformerFactory(configClient, 0)
+
+			cmLister := kubeInformer.Core().V1().ConfigMaps().Lister().ConfigMaps(defaults.ImageRegistryOperatorNamespace)
+			secretLister := kubeInformer.Core().V1().Secrets().Lister().Secrets(defaults.ImageRegistryOperatorNamespace)
+
+			proxyLister := configInformer.Config().V1().Proxies().Lister()
+
+			kubeInformer.Start(ctx.Done())
+			configInformer.Start(ctx.Done())
+
+			gd := &generatorDeployment{
+				driver:          &testDriver{},
+				coreClient:      kubeClient.CoreV1(),
+				proxyLister:     proxyLister,
+				cr:              &imageregistryv1.Config{},
+				configMapLister: cmLister,
+				secretLister:    secretLister,
+			}
+
+			cache.WaitForCacheSync(ctx.Done(), kubeInformer.Core().V1().ConfigMaps().Informer().HasSynced)
+			cache.WaitForCacheSync(ctx.Done(), kubeInformer.Core().V1().Secrets().Informer().HasSynced)
```

Comment:
Can we user a context with a timeout here? If these caches never get ready we won't know exactly what happen other the whole test timing out. Something like below would make it more clear (using a context with Timeout):

```
if !cache.WaitForCacheSync(
		ctx.Done(),
		kubeInformer.Core().V1().ConfigMaps().Informer().HasSynced,
		kubeInformer.Core().V1().Secrets().Informer().HasSynced,
	) {
		t.Fatal("caches not syncing")
	}
```